packages/babel-plugin-transform-object-rest-spread/src/index.js

```diff
@@ -21,24 +21,62 @@ export default function ({ types: t }) {
     return false;
   }
 
-  function createObjectSpread(file, props, objRef) {
-    const restElement = props.pop();
-
+  // extracts ["a", "b"] from {a: "hi", b}
+  function extractKeys(path) {
+    const propPaths = path.get("properties");
     const keys = [];
-    for (const prop of props) {
-      let key = prop.key;
-      if (t.isIdentifier(key) && !prop.computed) {
-        key = t.stringLiteral(prop.key.name);
+    for (const propPath of propPaths) {
+      let key = propPath.get("key").node;
+      if (t.isIdentifier(key) && !propPath.node.computed) {
+        key = t.stringLiteral(key.name);
       }
       keys.push(key);
     }
+    return keys;
+  }
+
+  function replaceImpureComputedKeys(path) {
+    const impureComputedPropertyDeclarators = [];
+    for (const propPath of path.get("properties")) {
+      const key = propPath.get("key");
+      if (propPath.node.computed && !key.isPure()) {
+        const identifier = path.scope.generateUidIdentifierBasedOnNode(key.node);
+        const declarator = t.variableDeclarator(identifier, key.node);
+        impureComputedPropertyDeclarators.push(declarator);
+        key.replaceWith(identifier);
+      }
+    }
+    return impureComputedPropertyDeclarators;
+  }
+
+  //expects path to an object pattern
+  function createObjectSpread(path, file, objRef) {
+    const last = path.get("properties")[path.get("properties").length - 1];
```

Comment:
For some reason I think this is the "proper" way - by using path method `.remove()`. popping would be direct manipulation and it would bypass all the logic which updates bindings, etc. Correct?